### PR TITLE
Workbench actions for Tree Fuzzy Search Toggle

### LINKED ISSUE #173866: Associate a keyboard shortcut with the new "Fuzzy Match" toggle button
### Edit

It looks like there was a misunderstanding when this issue was closed as not planned. The perceived request was "assign a default keyboard shortcut to toggle Fuzzy Match", but what I meant was "_just make it possible at all_ to assign a keyboard shortcut to toggle Fuzzy Match. As it stands, clicking the GUI element for toggling Fuzzy appears to be the only way to do it.

I'm not gonna spam a new issue about this minor nitpick again so soon. If someone stumbles upon this issue and wants to request the feature properly next time, be clearer with a title like, "Allow the user to assign a keyboard shortcut to toggle Fuzzy Match in various places", e.g. in the Explorer filter menu, the "Terminal: Recent command" and "Terminal: Recent directory", etc.

---

### Original comment

A "Fuzzy Match" toggle was introduced in https://github.com/microsoft/vscode/pull/164376, and is [documented in January 2023 release notes](https://code.visualstudio.com/updates/v1_75#_tree-find-contiguous-matching). Would it be possible to assign a keyboard shortcut to this toggle as well?

### PR DESCRIPTION
**Pull request by keerthivasansa**
Fixes #173866

- Associated Issue:  #173866 
- The code is up-to-date with the main branch

#### Changes
- Registered Workbench actions to toggle the Fuzzy Search behaviour in the Explorer Search Menu.
- The actions can be mapped to keyboard shortcuts hence allowing the user to create a keyboard shortcut to toggle or switch to a specific find match type.
- The registered actions are: 
      - Explorer: Toggle Find Match Type
      - Explorer: Switch to Fuzzy Find Match Type
      - Explorer: Switch to Contagious Find Match Type

#### Technical Changes
- Since the other commands in command palette are associated with focusing on views, no separate actions file was present for explorer category. Thus, I have created `explorerActions.ts` in `vs/base/browser/actions/explorerActions.ts` and registered the actions in it.
- A getter and setter for `findMatchType` was required in AsyncDataTree for this to be modified from an action. Async Data Tree was modified accordingly.

### CODE REVIEW

**Comment by keerthivasansa** [comment]
@microsoft-github-policy-service agree

File: src/vs/workbench/browser/actions/explorerActions.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	}
+}
+
+class SwitchToContigiousFindMatchType extends Action2 {
+	static readonly ID = 'workbench.action.contigiousFindMatchType';
+
+	constructor() {
+		super({
+			id: SwitchToContigiousFindMatchType.ID,
+			title: { value: localize('contigiousFindMatchType', 'Switch to Contigious Find Match Type'), original: 'Switch to Contigious Find Match Type' },
```
Comment (gjsjohnmurray) [inline src/vs/workbench/browser/actions/explorerActions.ts:67]:
```suggestion
			title: { value: localize('contiguousFindMatchType', 'Switch to Contiguous Find Match Type'), original: 'Switch to Contiguous Find Match Type' },
```

File: src/vs/workbench/browser/actions/explorerActions.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		const widget = accessor.get(IListService).lastFocusedList;
+		if (widget instanceof AbstractTree || widget instanceof AsyncDataTree) {
+			const tree = widget;
+			tree.findMatchType = TreeFindMatchType.Fuzzy;
+		}
+	}
+}
+
+class SwitchToContigiousFindMatchType extends Action2 {
+	static readonly ID = 'workbench.action.contigiousFindMatchType';
```
Comment (gjsjohnmurray) [inline src/vs/workbench/browser/actions/explorerActions.ts:62]:
```suggestion
	static readonly ID = 'workbench.action.contiguousFindMatchType';
```

File: src/vs/workbench/browser/actions/explorerActions.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		if (widget instanceof AbstractTree || widget instanceof AsyncDataTree) {
+			const tree = widget;
+			tree.findMatchType = TreeFindMatchType.Contiguous;
+		}
+	}
+}
+
+registerAction2(ToggleFindMatchType);
+registerAction2(SwitchToFuzzyFindMatchType);
+registerAction2(SwitchToContigiousFindMatchType);
```
Comment (gjsjohnmurray) [inline src/vs/workbench/browser/actions/explorerActions.ts:85]:
```suggestion
registerAction2(SwitchToContiguousFindMatchType);
```

File: src/vs/workbench/browser/actions/explorerActions.ts (hunk 2 above)
Comment (keerthivasansa) [reply src/vs/workbench/browser/actions/explorerActions.ts:62]: Wow, such a silly mistake 😅

File: src/vs/workbench/browser/actions/explorerActions.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		if (widget instanceof AbstractTree || widget instanceof AsyncDataTree) {
+			const tree = widget;
+			tree.findMatchType = TreeFindMatchType.Contiguous;
+		}
+	}
+}
+
+registerAction2(ToggleFindMatchType);
+registerAction2(SwitchToFuzzyFindMatchType);
+registerAction2(SwitchToContiguousFindMatchType);
```
Comment (joaomoreno) [inline src/vs/workbench/browser/actions/explorerActions.ts:85]: All these actions are generic enough to be general tree actions. Let's move them to `listCommands.ts` and change their category not to be Explorer specific.

File: src/vs/workbench/workbench.common.main.ts [hunk 5]
```diff
@@ -25,6 +25,7 @@ import 'vs/workbench/browser/actions/windowActions';
 import 'vs/workbench/browser/actions/workspaceActions';
 import 'vs/workbench/browser/actions/workspaceCommands';
 import 'vs/workbench/browser/actions/quickAccessActions';
+import 'vs/workbench/browser/actions/explorerActions';
```
Comment (joaomoreno) [inline src/vs/workbench/workbench.common.main.ts:28]: Let's remove this, once we move the commands to the right place.

**Review by joaomoreno — CHANGES_REQUESTED** [review]

**Review by keerthivasansa — COMMENTED** [review]
Requested changes made in the latest commit.

**Comment by keerthivasansa** [comment]
Hi @joaomoreno, I think it's ready for re-review. I have made the requested changes in https://github.com/microsoft/vscode/pull/175137/commits/3443042b91a0ef951a60dd2d8c536acfd63a0a2b
Please have a look and let me know if everything looks good or any other changes are required.

**Comment by keerthivasansa** [comment]
@joaomoreno Are there any pending changes I have to implement?

**Comment by keerthivasansa** [comment]
@gjsjohnmurray @joaomoreno Are there any pending changes that I have to implement to close this?

**Review by gjsjohnmurray — COMMENTED** [review]
This needs reviews by team members, and I am not one. Pinging @joaomoreno

**Comment by joaomoreno** [comment]
Thanks for your contribution! I have simplified it to become a simple toggle. 🍻

**Review by joaomoreno — APPROVED** [review]

**Review by lramos15 — APPROVED** [review]

**Pull request merged**